include/libdnf-cli/progressbar/progress_bar.hpp

```diff
@@ -0,0 +1,150 @@
+    READY,    // no color
+    STARTED,  // no color
+    SUCCESS,  // green
+    WARNING,  // yellow
+    ERROR,    // red
+};
+
+
+enum class MessageType : int {
+    INFO,     // no color
+    SUCCESS,  // green
+    WARNING,  // yellow
+    ERROR,    // red
+};
+
+
+
+
+
+class ProgressBar {
+public:
+    using Message = std::pair<MessageType, std::string>;
+
+    explicit ProgressBar(int64_t total_ticks);
+    explicit ProgressBar(int64_t total_ticks, const std::string & description);
+    virtual ~ProgressBar() = default;
+
+    // ticks
+    int64_t get_ticks() { return ticks; }
+    void set_ticks(int64_t value);
+    void add_ticks(int64_t value);
+
+    // total ticks
+    int64_t get_total_ticks() { return total_ticks; }
+    void set_total_ticks(int64_t value);
+    //void add_total_ticks(int64_t value);
+
+    // number
+    int32_t get_number() { return number; }
+    void set_number(int32_t value) { number = value; }
+
+    // total
+    int32_t get_total() { return total; }
+    void set_total(int32_t value) { total = value; }
+
+    // workflow
+    void start();
+    ProgressBarState get_state() { return state; }
+    void set_state(ProgressBarState value) { update(); state = value; }
+    bool is_finished() { return get_state() != ProgressBarState::READY && get_state() != ProgressBarState::STARTED; }
+    bool is_failed() { return get_state() == ProgressBarState::ERROR || get_state() == ProgressBarState::WARNING; }
+
+    // description
+    std::string get_description() const { return description; }
+    void set_description(const std::string & value) { description = value; }
+
+    // messages
+    void add_message(MessageType type, const std::string & message) { messages.push_back({type, message}); }
+    const std::vector<Message> & get_messages() { return messages; }
+
+    // auto-finish feature; turn off if you want to handle state manually
+    bool get_auto_finish() { return auto_finish; }
+    void set_auto_finish(bool value) { auto_finish = value; }
+
+    // stats
+    void update();
+    int32_t get_percent_done() { return percent_done; }
+    int64_t get_current_speed() { return current_speed; }
+    int64_t get_average_speed() { return average_speed; }
+    int64_t get_elapsed_seconds() { return elapsed_seconds; }
+    int64_t get_remaining_seconds() { return remaining_seconds; }
+
+    std::chrono::time_point<std::chrono::system_clock> get_begin() { return begin; }
+
+    friend std::ostream & operator<<(std::ostream & os, ProgressBar & bar);
+
+protected:
+    virtual void to_stream(std::ostream & stream) = 0;
+
+private:
+    // ticks
+    int64_t ticks = 0;
+    int64_t total_ticks = 0;
+
+    // numbers
+    int32_t number = 0;
+    int32_t total = 0;
+
+    // description
+    std::string description;
+
+    // messages
+    std::vector<Message> messages;
+
+    ProgressBarState state = ProgressBarState::READY;
+
+    int32_t percent_done = 0;
+    int64_t elapsed_seconds = 0;
```

Comment:
Wow. `int64_t` for seconds. We can express the age of the Earth.

Reply:
Right :)
In fact, it's there most likely to use a compatible type with: `elapsed_seconds = std::chrono::duration_cast<std::chrono::seconds>(delta).count()`